src/Processors/QueryPlan/QueryPlan.cpp

```diff
@@ -180,6 +185,8 @@ QueryPipelinePtr QueryPlan::buildQueryPipeline(
     for (auto & context : interpreter_context)
         last_pipeline->addInterpreterContext(std::move(context));
 
+    if (max_parallel_streams > 0)
+        last_pipeline->setMaxParallelStreams(max_parallel_streams);
```

Comment:
It is just a sum for all leaves. It is not correct to use in many cases (e.g. sometimes we want to read-in-order from a lot of sources, but it cannot be efficiently done in more then one thread).